### PR TITLE
Optionally check for indexing outside of map array bounds

### LINKED ISSUE #175: Exposing kernel pointers to user mode apps via map keys
Same as issue #170 except that 170 did so via map values whereas this issue is about map keys.
Sample program is in https://github.com/vbpf/ebpf-samples/pull/4

A closely related issue is that a program can use an array index that is out of bounds.
For an array map, the verifier should really validate that the key is a number in the range [0, max_entries-1].

### PR DESCRIPTION
**Pull request by dthaler**
Provide a "strict" option that is not the default.  If the developer opts in to this flag, the verifier will do some additional checks for things that are potential bugs in the program.   Per comments in #175 from poornagmsft and xfoukas, this can help with preventing failures, improving developer efficient, and even improving program performance compared to other alternatives.

Fixes #175

Signed-off-by: Dave Thaler <dthaler@ntdev.microsoft.com>

### CODE REVIEW

**Comment by elazarg** [comment]
Regarding the description: the `strict` flag should not be about potential (arbitrary) bugs in the program. It is about potential speedups, not covered by current specification. In other words, this flag change the specification of array maps to disallow out-of-bounds access, thus allowing the trusted environment to do away with some runtime checks.

It's the same thing in this PR; I'm just noting this here again to make this design decision explicit.

File: src/crab/ebpf_domain.hpp [hunk 1]
```diff
@@ -521,6 +527,21 @@ class ebpf_domain_t final {
         if (!when_stack.is_bottom()) {
             if (!stack.all_num(when_stack, lb, ub)) {
                 require(when_stack, access_reg.type != T_STACK, "Illegal map update with a non-numerical value.");
+            } else if (thread_local_options.strict && map_type.has_value() && map_type->is_array) {
+                // Get offset value.
+                variable_t key_ptr = access_reg.offset;
+                std::optional<number_t> offset = m_inv[key_ptr].singleton();
+                if (!offset.has_value()) {
+                    // REVIEW: is there a better way to fail validation here?
+                    require(m_inv, linear_expression_t(0) != 0, "Pointer must be a singleton");
```
Comment (elazarg) [inline src/crab/ebpf_domain.hpp:536]: There's no better way that I know of right now. We should probably have an `require_false()` method encapsulating this.

File: src/crab/ebpf_domain.hpp [hunk 2]
```diff
@@ -521,6 +527,21 @@ class ebpf_domain_t final {
...
+                } else if (s.key) {
+                    // Look up the value pointed to by the key pointer.
+                    variable_t key_value =
+                        variable_t::cell_var(data_kind_t::values, (uint64_t)offset.value(), sizeof(uint32_t));
```
Comment (elazarg) [inline src/crab/ebpf_domain.hpp:540]: This too might be useful more generally so it merits a method, though I don't think it's needed anywhere else right now.

**Review by elazarg — APPROVED** [review]
LGTM. I guess you'd want to remove the `REVIEW` comment? Additionally, let me know if you want to address the other note too, or to merge as it is.

File: src/crab/ebpf_domain.hpp (hunk 1 above)
Comment (dthaler) [reply src/crab/ebpf_domain.hpp]: done

**Comment by dthaler** [comment]
> LGTM. I guess you'd want to remove the `REVIEW` comment? Additionally, let me know if you want to address the other note too, or to merge as it is.

Removed the REVIEW comment and did the require_false suggestion.  For the other note I think it would be better to wait until it's needed somewhere else, since the prototype of whatever method is used to put the code in might be quite different depending on the second caller's needs.  Unlike require_false() it's hard for me to predict what it would need to look like to be more generally useful.

**Pull request merged**